
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Learn:利用shell和perl构建转录组(RNA-seq)分析流程</title>
    <meta name="description" content="Mix together to build a RNA-seq pipeline.">
    <meta name="author" content="Zhoujj">

    <!-- Enable responsive viewport -->
    <meta name="viewport" content="width=device-width, initial-scale=1.0">

    <!-- Le HTML5 shim, for IE6-8 support of HTML elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <!-- Le styles -->
    <link href="/assets/themes/twitter/bootstrap/css/bootstrap.2.2.2.min.css" rel="stylesheet">
    <link href="/assets/themes/twitter/css/style.css?body=1" rel="stylesheet" type="text/css" media="all">
    
    <!--syntax-->
	<link href="/assets/themes/twitter/css/syntax.css" rel="stylesheet">
    <!-- Le fav and touch icons -->
  <!-- Update these with your own images
    <link rel="shortcut icon" href="images/favicon.ico">
    <link rel="apple-touch-icon" href="images/apple-touch-icon.png">
    <link rel="apple-touch-icon" sizes="72x72" href="images/apple-touch-icon-72x72.png">
    <link rel="apple-touch-icon" sizes="114x114" href="images/apple-touch-icon-114x114.png">
  -->

    <!-- atom & rss feed -->
    <link href="/atom.xml" type="application/atom+xml" rel="alternate" title="Sitewide ATOM Feed">
    <link href="/rss.xml" type="application/rss+xml" rel="alternate" title="Sitewide RSS Feed">

  </head>

  <body>
    <div class="navbar">
      <div class="navbar-inner">
        <div class="container-narrow">
        <!--<a class="brand" href="/">UnCoverIt</a>-->
          <ul class="nav">
			<li><a href="/">Home</a></li>
            
            
            


  
    
      
      	
      	<li><a href="/about.html">About</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/categories.html">Categories</a></li>
      	
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/tags.html">Tags</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/archive.html">Archive</a></li>
      	
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/pages.html">Pages</a></li>
      	
      
    
  
    
      
    
  
    
      
    
  



          </ul>
        </div>
      </div>
    </div>

    <div class="container-narrow">

      <div class="content">
        
<div class="page-header">
  <h1>Learn:利用shell和perl构建转录组(RNA-seq)分析流程 </h1>
</div>

<div class="row-fluid post-full">
  <div class="span12">
    <div class="date">
      <span>19 July 2014 posted by <a href="http://zhoujj2013.github.io/~zhoujj/">zhoujj</a></span>
    </div>
    <div class="content">
      
<p>随着NGS的高速发展，越来越多有价值的物种被测序，随之而来的，基因表达及表观遗传数据会急剧增加。</p>

<p>公共数据库有大量已经公开的数据，这些数据大多数是基于二代测序技术产生的。</p>

<p>fly/C.elegan:</p>

<p><a href="http://www.modencode.org/">http://www.modencode.org/</a></p>

<p>Mouse and Human:</p>

<p><a href="http://genome.ucsc.edu/ENCODE/">http://genome.ucsc.edu/ENCODE/</a></p>

<p>GEO/ENA</p>

<p><a href="http://www.ncbi.nlm.nih.gov/geo/">http://www.ncbi.nlm.nih.gov/geo/</a></p>

<p><a href="http://www.ebi.ac.uk/ena/">http://www.ebi.ac.uk/ena/</a></p>

<h4 id="zebra-fish">zebra fish转录分析方案设计</h4>

<p><strong>生物学问题</strong>: 比较同一时期，male body 和 male head样品的基因表达的差异。</p>

<p><strong>数据来源</strong>: 参考文章，<a href="http://genome.cshlp.org/content/22/10/2067.long">http://genome.cshlp.org/content/22/10/2067.long</a></p>

<p>male_body(ERR004012): </p>

<p>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR004/ERR004012/ERR004012_1.fastq.gz</p>

<p>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR004/ERR004012/ERR004012_2.fastq.gz</p>

<p>male_head(ERR004022): </p>

<p>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR004/ERR004022/ERR004022_1.fastq.gz</p>

<p>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR004/ERR004022/ERR004022_2.fastq.gz</p>

<p><strong>方案设计</strong>:</p>

<p><img src="https://dl.dropboxusercontent.com/u/87531643/web/rna-seq.png" alt="rna-seq_diagram" title="数据分析流程图" /></p>

<h4 id="section">测序数据质量控制</h4>

<p>下载测序数据：</p>

<div class="highlight"><pre><code class="bash">mkdir ZabraRnaSeq
<span class="nb">cd </span>ZabraRnaSeq
mkdir data
<span class="nb">cd </span>data
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR004/ERR004012/ERR004012_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR004/ERR004012/ERR004012_2.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR004/ERR004022/ERR004022_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR004/ERR004022/ERR004022_2.fastq.gz
</code></pre></div>

<p>用fastqc对数据质量进行查看：</p>

<div class="highlight"><pre><code class="bash"><span class="c"># $fastqc = path_to_fastqc_dir</span>
ls *.gz | xargs gunzip
<span class="nv">$fastqc</span>/fastqc stdin -o ./ ./ERR004012_1.fastq
<span class="nv">$fastqc</span>/fastqc stdin -o ./ ./ERR004012_2.fastq
<span class="nv">$fastqc</span>/fastqc stdin -o ./ ./ERR004022_1.fastq
<span class="nv">$fastqc</span>/fastqc stdin -o ./ ./ERR004022_2.fastq
</code></pre></div>

<p>通过fastqc结果判断数据是否合格，具体判断标准,请参考下面链接的Example Reports部分：</p>

<p><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">http://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></p>

<p>如果你发现有些数据不合格，或者还有adapter，应该用相应的程序把低质量数据去除。</p>

<p>如果你对fastq文件数据质量体系没有概念，请先阅读<a href="http://en.wikipedia.org/wiki/FASTQ_format">fastq文件格式</a>和<a href="http://en.wikipedia.org/wiki/Phred_quality_score">Phred_quality_score</a>:</p>

<p>用fastx去除低质量数据：</p>

<div class="highlight"><pre><code class="bash"><span class="c"># $fastx = path_to_fastx_dir</span>
<span class="c"># GA2 phred 33, read length=75</span>
<span class="nv">$fastx</span>/bin/fastq_quality_trimmer -t 36 -l 35 -i ./ERR004012_1.fastq -o ./ERR004012_1.trimed.fq
<span class="nv">$fastx</span>/bin/fastq_quality_trimmer -t 36 -l 35 -i ./ERR004012_1.fastq -o ./ERR004012_2.trimed.fq
<span class="nv">$fastx</span>/bin/fastq_quality_trimmer -t 36 -l 35 -i ./ERR004012_1.fastq -o ./ERR004022_1.trimed.fq
<span class="nv">$fastx</span>/bin/fastq_quality_trimmer -t 36 -l 35 -i ./ERR004012_1.fastq -o ./ERR004022_2.trimed.fq
<span class="nb">cd</span> .. <span class="c"># 返回原来的目录</span>
</code></pre></div>

<p>此处没有出现高频adapter序列,所以就不用跑$fastx/bin/fastx_clipper。</p>

<h4 id="reads">短reads比对及结果分析</h4>

<p>准备参考序列及参考序列index:</p>

<div class="highlight"><pre><code class="bash">mkdir Zebra <span class="o">&amp;&amp;</span> <span class="nb">cd </span>Zebra
<span class="k">for </span>i in 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 MT; <span class="k">do </span>wget ftp://ftp.ensembl.org/pub/release-75/fasta/danio_rerio/dna/Danio_rerio.Zv9.75.dna.chromosome.<span class="nv">$i</span>.fa.gz;done;
zcat *.fa.gz | sed <span class="s1">&#39;s/^/chr/g&#39;</span> &gt; Danio_rerio.Zv9.75.Chr.fa
rm *.fa.gz
bowtie2-build ./Danio_rerio.Zv9.75.Chr.fa ./Danio_rerio.Zv9.75.Chr

<span class="c"># download geneset</span>
wget ftp://ftp.ensembl.org/pub/release-75/gtf/danio_rerio/Danio_rerio.Zv9.75.gtf.gz
<span class="k">for </span>i in 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 MT; <span class="k">do </span>zcat Danio_rerio.Zv9.75.gtf.gz | awk <span class="s1">&#39;$1 == $i&#39;</span> | sed <span class="s1">&#39;s/^/chr/g&#39;</span> &gt;&gt; Danio_rerio.Zv9.75.gtf;done;
rm Danio_rerio.Zv9.75.gtf.gz
<span class="nb">cd</span> .. <span class="c"># 返回原来的目录</span>
</code></pre></div>

<p>用tophat比对到zebra fish参考基因组：</p>

<div class="highlight"><pre><code class="bash"><span class="c"># $tophat = path_to_tophat_dir</span>
<span class="c"># 注意查看内存，两个taphat很容易把内存消耗掉的</span>
nohup <span class="nv">$tophat</span>/tophat2 -G ./Zebra/Danio_rerio.Zv9.75.gtf -o male_body_tophat -p 8 ./Zebra/Danio_rerio.Zv9.75.Chr ./data/ERR004012_1.trimed.fq ./ERR004012_2.trimed.fq &gt;MB.tophat.std 2&gt;MB.tophat.err&amp;
nohup <span class="nv">$tophat</span>/tophat2 -G ./Zebra/Danio_rerio.Zv9.75.gtf -o male_head_tophat -p 8 ./Zebra/Danio_rerio.Zv9.75.Chr ./data/ERR004022_1.trimed.fq ./ERR004022_2.trimed.fq &gt;MH.tophat.std 2&gt;MH.tophat.err&amp;
<span class="c"># 跑完后请从MB.tophat.err，MH.tophat.err查看基本比对结果统计。</span>
<span class="c"># 具体结果文件，请查看taphat的文档</span>
<span class="c"># 主要结果文件是，accepted_hits.bam文件</span>
</code></pre></div>

<h4 id="section-1">表达量计算及差异表达分析</h4>

<p>用cufflinks组装新的转录集合及计算每一个转录本的表达量：</p>

<div class="highlight"><pre><code class="bash"><span class="c"># $cufflinks = path_to_cufflinks_dir</span>
<span class="nv">$cufflinks</span>/cufflinks -g ./Zebra/Danio_rerio.Zv9.75.gtf -o ./male_body_cufflinks ./male_body_tophat/accepted_hits.bam &gt;MB.cufflinks.std 2&gt;MB.cufflnks.err&amp;
<span class="nv">$cufflinks</span>/cufflinks -g ./Zebra/Danio_rerio.Zv9.75.gtf -o ./male_head_cufflinks ./male_head_tophat/accepted_hits.bam &gt;MH.cufflinks.std 2&gt;MH.cufflnks.err&amp;
</code></pre></div>

<p>合并两个样品的转录本：</p>

<div class="highlight"><pre><code class="bash">vim transcript.lst
<span class="c"># 加入transcript的路径</span>
./male_body_cufflinks/transcripts.gtf
./male_head_cufflinks/transcripts.gtf

mkdir ./merged_asm
<span class="nv">$cufflinks</span>/cuffmerge -o ./merged_asm -g ./Zebra/Danio_rerio.Zv9.75.gtf -s ./Zebra/Danio_rerio.Zv9.75.Chr.fa -p 8 ./transcript.lst &gt; cuffmerge.std 2&gt; cuffmerge.err &amp;
</code></pre></div>

<p>用cuffdiff比较两个样品的差异表达基因：</p>

<div class="highlight"><pre><code class="bash"><span class="nv">$cufflinks</span>/cuffdiff -o MBvsMH_cuffdiff -L MB,MH -p 8 ./merged_asm/transcripts.gtf ./male_body_tophat/accepted_hits.bam ./male_head_tophat/accepted_hits.bam &gt;MBvsMH_cuffdiff.std 2&gt;MBvsMH_cuffdiff.err &amp;
</code></pre></div>

<h4 id="section-2">生物信息分析报告</h4>

<p>可以根据不同的需要组织数据分析报告，主要以下几点：</p>

<ol>
  <li>
    <p>测序数据基本统计及质量评估；</p>
  </li>
  <li>
    <p>短序列比对结果统计；</p>
  </li>
  <li>
    <p>转录本数目、表达量统计；</p>
  </li>
  <li>
    <p>样品间基因表达差异分析结果；</p>
  </li>
  <li>
    <p>进行必要的基因功能注释分析。</p>
  </li>
</ol>

<p>上面把RNA-seq数据分析实现了一次，有人可能会问，你这也可以算是流程吗？</p>

<p>分析流程本来就是这样的，你只需要用shell或者perl把上面的每一步串联起来，就是一个自动化的流程了。</p>

<p>转载请标明出处：</p>

<p><a href="http://zhoujj2013.github.io/UnCoverIt/bioinformatics/2014/07/19/bioinfo-mix-together/">http://zhoujj2013.github.io/UnCoverIt/bioinformatics/2014/07/19/bioinfo-mix-together/</a></p>

    </div>

  
    <ul class="tag_box inline">
      <li><i class="icon-folder-open"></i></li>
      
      


  
     
    	<li><a href="/categories.html#bioinformatics-ref">
    		bioinformatics <span>10</span>
    	</a></li>
    
  


    </ul>
    

  
    <ul class="tag_box inline">
      <li><i class="icon-tags"></i></li>
      
      


  
     
    	<li><a href="/tags.html#bioinformatics-ref">bioinformatics <span>10</span></a></li>
     
    	<li><a href="/tags.html#RnaSeq-ref">RnaSeq <span>1</span></a></li>
     
    	<li><a href="/tags.html#NGS-ref">NGS <span>1</span></a></li>
     
    	<li><a href="/tags.html#learn-ref">learn <span>5</span></a></li>
    
  



    </ul>
    

    <hr>
    <div class="pagination">
      <ul>
      
        <li class="prev"><a href="/bioinformatics/2014/07/18/bioinfo-tools-collect-and-use" title="Learn:常用生物信息分析软件<i>陈列馆</i>">&larr; Previous</a></li>
      
        <li><a href="/archive.html">Archive</a></li>
      
        <li class="next disabled"><a>Next &rarr;</a>
      
      </ul>
    </div>
    <hr>
    


  <div id="disqus_thread"></div>
<script type="text/javascript">
    var disqus_developer = 1;
    var disqus_shortname = 'zhoujj'; // required: replace example with your forum shortname
    
    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = 'http://' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="http://disqus.com" class="dsq-brlink">blog comments powered by <span class="logo-disqus">Disqus</span></a>




  </div>
</div>


      </div>
      <hr>
      <footer>
        <p>Copyright &copy; 2014 <a href="http://zhoujj2013.github.io/~zhoujj/index.html" title="Zhoujj's homepage" target="_blank">Zhoujj</a> 
          with help from <a href="http://jekyllbootstrap.com" target="_blank" title="The Definitive Jekyll Blogging Framework">Jekyll Bootstrap</a>
          and <a href="http://twitter.github.com/bootstrap/" target="_blank">Twitter Bootstrap</a>
        </p>
      </footer>

    </div>

    
  </body>
</html>

