
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>JBrowser:导入vcf wig和bam数据</title>
    <meta name="description" content="Import vcf, wig and bam to jbrowser.">
    <meta name="author" content="Zhoujj">

    <!-- Enable responsive viewport -->
    <meta name="viewport" content="width=device-width, initial-scale=1.0">

    <!-- Le HTML5 shim, for IE6-8 support of HTML elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <!-- Le styles -->
    <link href="/assets/themes/twitter/bootstrap/css/bootstrap.2.2.2.min.css" rel="stylesheet">
    <link href="/assets/themes/twitter/css/style.css?body=1" rel="stylesheet" type="text/css" media="all">

    <!-- Le fav and touch icons -->
  <!-- Update these with your own images
    <link rel="shortcut icon" href="images/favicon.ico">
    <link rel="apple-touch-icon" href="images/apple-touch-icon.png">
    <link rel="apple-touch-icon" sizes="72x72" href="images/apple-touch-icon-72x72.png">
    <link rel="apple-touch-icon" sizes="114x114" href="images/apple-touch-icon-114x114.png">
  -->

    <!-- atom & rss feed -->
    <link href="/atom.xml" type="application/atom+xml" rel="alternate" title="Sitewide ATOM Feed">
    <link href="/rss.xml" type="application/rss+xml" rel="alternate" title="Sitewide RSS Feed">

  </head>

  <body>
    <div class="navbar">
      <div class="navbar-inner">
        <div class="container-narrow">
        <!--<a class="brand" href="/">UnCoverIt</a>-->
          <ul class="nav">
			<li><a href="/">Home</a></li>
            
            
            


  
    
      
      	
      	<li><a href="/about.html">About</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/categories.html">Categories</a></li>
      	
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/tags.html">Tags</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/archive.html">Archive</a></li>
      	
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/pages.html">Pages</a></li>
      	
      
    
  
    
      
    
  
    
      
    
  



          </ul>
        </div>
      </div>
    </div>

    <div class="container-narrow">

      <div class="content">
        
<div class="page-header">
  <h1>JBrowser:导入vcf wig和bam数据 </h1>
</div>

<div class="row-fluid post-full">
  <div class="span12">
    <div class="date">
      <span>16 July 2014 posted by <a href="http://zhoujj2013.github.io/~zhoujj/">zhoujj</a></span>
    </div>
    <div class="content">
      
<h4 id="vcf-wigbam">vcf, wig和bam文件的介绍</h4>

<p>vcf文件主要是用来记录基因组variants位置及variants功能注释。</p>

<p>详细格式请查看：</p>

<p>http://samtools.github.io/hts-specs/</p>

<p>wig是用来表示连续和非连续数据信号的文件格式。</p>

<p>详细格式请查看：</p>

<p>http://www.genome.ucsc.edu/goldenPath/help/wiggle.html</p>

<p>bam是短片断序列alignment结果sam文件，通过samtools排序压缩后生成的文件，用于存储比对文件。
bam文件需要通过samtools进行查看。</p>

<p>详细格式请查看：</p>

<p>http://samtools.github.io/hts-specs/SAMv1.pdf</p>

<p>http://genome.sph.umich.edu/wiki/SAM</p>

<p>http://picard.sourceforge.net/explain-flags.html</p>

<p>http://wiki.bits.vib.be/index.php/.sam</p>

<div class="highlight"><pre><code class="bash">samtool view &lt;bam_file&gt;
</code></pre></div>

<h4 id="vcf">导入vcf文件</h4>

<p>我个人觉得tracks.conf配置文件的配置内容更加易读，所以接下来的配置信息，我都写到tracks.conf文件。</p>

<p>vcf文件有两种显示方式：CanvasVariants和HTMLVariants。我暂没有发现两者的区别。</p>

<p>在导入vcf文件前要对vcf文件进行压缩，并建立index，这个过程要求安装samtools/tabix。</p>

<div class="highlight"><pre><code class="bash"><span class="c"># tabix: https://github.com/samtools/tabix</span>
<span class="c"># 下载并安装tabix</span>
git clone https://github.com/samtools/tabix.git
<span class="nb">cd </span>tabix
make

<span class="c"># 对vcf进行压缩,以s410.vcf为例</span>
<span class="nv">$tabix</span>/bgzip s410.vcf

<span class="c"># 建立index</span>
<span class="nv">$tabix</span>/tabix -p s410.vcf.gz

<span class="c"># 生成s410.vcf.gz.tbi</span>
</code></pre></div>

<p><strong>CanvasVariants</strong></p>

<div class="highlight"><pre><code class="bash"><span class="o">[</span> tracks . s410 <span class="o">]</span>
<span class="nv">storeClass</span> <span class="o">=</span> JBrowse/Store/SeqFeature/VCFTabix
<span class="nv">urlTemplate</span> <span class="o">=</span> ../../vcf/s410.vcf.gz
<span class="nv">category</span> <span class="o">=</span> VCF
<span class="nb">type</span> <span class="o">=</span> JBrowse/View/Track/CanvasVariants
<span class="nv">key</span> <span class="o">=</span> s410 VCF1
</code></pre></div>

<p><strong>HTMLVariants</strong></p>

<div class="highlight"><pre><code class="bash"><span class="o">[</span> tracks . s410-2 <span class="o">]</span>
hooks.modify <span class="o">=</span> <span class="k">function</span><span class="o">(</span> track, feature, div <span class="o">)</span> <span class="o">{</span> div.style.backgroundColor <span class="o">=</span> track.config.variantIsHeterozygous<span class="o">(</span>feature<span class="o">)</span> ? <span class="s1">&#39;red&#39;</span> : <span class="s1">&#39;blue&#39;</span>; <span class="o">}</span>
<span class="nv">key</span> <span class="o">=</span> s410 VCF2
<span class="c"># 区分杂合和纯合位点</span>
<span class="nv">variantIsHeterozygous</span> <span class="o">=</span> <span class="k">function</span><span class="o">(</span> feature <span class="o">)</span> <span class="o">{</span> var <span class="nv">genotypes</span> <span class="o">=</span> feature.get<span class="o">(</span><span class="s1">&#39;genotypes&#39;</span><span class="o">)</span>;  <span class="k">for</span><span class="o">(</span> var sampleName in genotypes <span class="o">)</span> <span class="o">{</span> try <span class="o">{</span> var <span class="nv">gtString</span> <span class="o">=</span> genotypes<span class="o">[</span>sampleName<span class="o">]</span>.GT.values<span class="o">[</span>0<span class="o">]</span>; <span class="k">if</span><span class="o">(</span> ! /^1<span class="o">([</span><span class="se">\|\/</span><span class="o">]</span>1<span class="o">)</span>*<span class="nv">$/</span>.test<span class="o">(</span> gtString<span class="o">)</span> <span class="o">&amp;&amp;</span> ! /^0<span class="o">([</span><span class="se">\|\/</span><span class="o">]</span>0<span class="o">)</span>*<span class="nv">$/</span>.test<span class="o">(</span> gtString <span class="o">)</span> <span class="o">)</span> <span class="k">return </span><span class="nb">true</span>; <span class="o">}</span> catch<span class="o">(</span>e<span class="o">)</span> <span class="o">{}</span> <span class="o">}</span> <span class="k">return </span><span class="nb">false</span>; <span class="o">}</span>
<span class="nv">storeClass</span> <span class="o">=</span> JBrowse/Store/SeqFeature/VCFTabix
<span class="nv">urlTemplate</span> <span class="o">=</span> ../../vcf/s410.vcf.gz
<span class="nb">type</span> <span class="o">=</span> JBrowse/View/Track/HTMLVariants
metadata.category <span class="o">=</span> VCF
metadata.Description <span class="o">=</span> Variants called from XX.bam using samtools and bcftools.  Heterozygous variants are shown in red, homozygous variants in blue.
</code></pre></div>

<h4 id="bam">导入bam文件</h4>

<p>bam文件导入后，有两种显示方式</p>

<ul>
  <li>
    <p>直接显示每一个reads的比对情况(type = JBrowse/View/Track/Alignments2)</p>
  </li>
  <li>
    <p>只显示比对的coverage情况(type = JBrowse/View/Track/SNPCoverage)</p>
  </li>
</ul>

<p>两种方式均要临时读进某个region的alignment数据，所显示的速度较为慢的。</p>

<p><strong>alignment2方式</strong></p>

<div class="highlight"><pre><code class="html">[ tracks . s410_bam ]                     #　以s410.bam为例，要求已经排序并建立index, **.bam.bai文件
style.height = 7
key = BAM - s410 alignment2
storeClass = JBrowse/Store/SeqFeature/BAM
urlTemplate = ../../data/s410.sorted.rmdup.bam
maxFeatureScreenDensity = 4
metadata.category = BAM
metadata.Description = small test pattern of BAM-format paired alignments of simulated resequencing reads on the s410 sample.
type = JBrowse/View/Track/Alignments2
chunkSizeLimit = 50000000                  # 这个是限制在某一个region允许读取的数据量，如果读进的数据量太大会造成浏览器“罢工”。
</code></pre></div>

<p><strong>coverage方式</strong></p>

<div class="highlight"><pre><code class="html">[ tracks . volvox-sorted_bam_coverage ]
storeClass = JBrowse/Store/SeqFeature/BAM
urlTemplate = ../../data/s410.sorted.rmdup.bam
metadata.category = BAM
metadata.Description = SNP/Coverage view of s410.bam, simulated resequencing alignments.
type = JBrowse/View/Track/SNPCoverage       #注意这个设置的区别
key = BAM - s410 SNPs/Coverage
chunkSizeLimit = 50000000
</code></pre></div>

<h4 id="wig">导入wig文件</h4>

<p>JBrowser并不直接支持wig文件，它要求wig文件压缩成bigwig格式，这个可以大大节约存储空间。</p>

<p>目前还没有现成的程序支持bam文件转换为wig文件，wig文件主要是用来看sequencing的coverage和density分布情况，所以先解决bam和wig文件转换问题。</p>

<p><strong>from bam file to wig file</strong></p>

<div class="highlight"><pre><code class="bash"><span class="c"># 先用samtools把bam转化为sam数据流，再通过管道把数据导入到sam2wig.pl脚本，生成wig文件</span>
samtools view ../data/s410.sorted.rmdup.bam | perl sam2wig.pl --bin 30 --minMQ 30 - &gt; s410.wig

<span class="c"># 下载ucsc tools</span>
wget http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/fetchChromSizes
wget http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/wigToBigWig

<span class="c"># 把wig转化为bigwig格式</span>
./fetchChromSizes hg19 &gt; hg19.chrom.sizes
./wigToBigWig ./s410.wig hg19.chrom.sizes s410.bw
</code></pre></div>

<p>sam2wig.pl代码如下：</p>

<script src="https://gist.github.com/zhoujj2013/2f44ba0c0437f0241cc4.js"></script>

<p><strong>配置bigwig文件</strong></p>

<p>bigwig有两种显示的方式，一种是density，另外一种是XY-plot。</p>

<ul>
  <li>density plot方式</li>
</ul>

<div class="highlight"><pre><code class="html">[ tracks . s410_bw_density ]
style.neg_color = function(feature) { return feature.get(&#39;score&#39;) <span class="nt">&lt; 150</span> <span class="err">?</span> <span class="err">&#39;</span><span class="na">green</span><span class="err">&#39;</span> <span class="na">:</span> <span class="err">&#39;</span><span class="na">red</span><span class="err">&#39;;</span> <span class="err">}</span> <span class="err">#</span> <span class="err">如果位点的</span><span class="na">sequencing</span> <span class="na">depth</span> <span class="nt">&gt;</span> 150X显示为红色，否则绿色
bicolor_pivot = mean
key = s410 BigWig Density
storeClass = JBrowse/Store/BigWig
urlTemplate = ../../sam2wig/s410
type = JBrowse/View/Track/Wiggle/Density
metadata.category = Quantitative / Density # 这里是表示两层的分类目录，用/分开。
metadata.Description = Wiggle/Density view of s410.bw.  Also demonstrates use of a user-configurable callback to set the value of neg_color to green when the score is below 150.
</code></pre></div>

<ul>
  <li>XY-plot方式</li>
</ul>

<div class="highlight"><pre><code class="html">[ tracks . s410_bw_xyplot ]
style.pos_color = function(feature) { return feature.get(&#39;score&#39;) &gt; 30 ? &#39;red&#39; : &#39;blue&#39;; }
variance_band = true
key = s410 BigWig XY
storeClass = JBrowse/Store/BigWig
urlTemplate = ../../sam2wig/s410.bw
type = JBrowse/View/Track/Wiggle/XYPlot
metadata.category = Quantitative / XY Plot
metadata.description = Wiggle/XYPlot view of volvox_microarray.bw.  Demonstrates use of a user-configured callback to set the bar color to red when the score is above 30.
</code></pre></div>

<p>完成vcf、wig、bam文件导入后，不需要建立索引，直接就可以查看信号。</p>

<p>参考：
http://gmod.org/wiki/JBrowse_Configuration_Guide</p>

<p>转载请标明出处：
http://zhoujj2013.github.io/UnCoverIt/bioinformatics/2014/07/16/jbrowser-import-vcf-wig-bam/</p>


    </div>

  
    <ul class="tag_box inline">
      <li><i class="icon-folder-open"></i></li>
      
      


  
     
    	<li><a href="/categories.html#bioinformatics-ref">
    		bioinformatics <span>10</span>
    	</a></li>
    
  


    </ul>
    

  
    <ul class="tag_box inline">
      <li><i class="icon-tags"></i></li>
      
      


  
     
    	<li><a href="/tags.html#bioinformatics-ref">bioinformatics <span>10</span></a></li>
     
    	<li><a href="/tags.html#visualization-ref">visualization <span>5</span></a></li>
     
    	<li><a href="/tags.html#web-ref">web <span>5</span></a></li>
    
  



    </ul>
    

    <hr>
    <div class="pagination">
      <ul>
      
        <li class="prev"><a href="/bioinformatics/2014/07/16/jbrowser-import-ucsc-tracks" title="JBrowser:导入UCSC数据库基因信息数据">&larr; Previous</a></li>
      
        <li><a href="/archive.html">Archive</a></li>
      
        <li class="next"><a href="/bioinformatics/2014/07/16/simple-shell-command-for-bioinfo" title="Learn:用shell命令完成简单生物信息分析查询及统计">Next &rarr;</a></li>
      
      </ul>
    </div>
    <hr>
    


  <div id="disqus_thread"></div>
<script type="text/javascript">
    var disqus_developer = 1;
    var disqus_shortname = 'zhoujj'; // required: replace example with your forum shortname
    
    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = 'http://' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="http://disqus.com" class="dsq-brlink">blog comments powered by <span class="logo-disqus">Disqus</span></a>




  </div>
</div>


      </div>
      <hr>
      <footer>
        <p>Copyright &copy; 2014 <a href="http://zhoujj2013.github.io/~zhoujj/index.html" title="Zhoujj's homepage" target="_blank">Zhoujj</a> 
          with help from <a href="http://jekyllbootstrap.com" target="_blank" title="The Definitive Jekyll Blogging Framework">Jekyll Bootstrap</a>
          and <a href="http://twitter.github.com/bootstrap/" target="_blank">Twitter Bootstrap</a>
        </p>
      </footer>

    </div>

    
  </body>
</html>

